src/lib.rs

```diff
@@ -0,0 +1,155 @@
+                if char != '-' {
+                    return false;
+                }
+            }
+            _ => {
+                match char {
+                    '0' ... '9' | 'a' ... 'f' => (),
+                    _ => return false,
+                }
+            }
+        }
+    }
+    return true;
+}
+
+/// Checks whether its argument is a blacklist file that's usable in the algorithm at
+/// https://webbluetoothcg.github.io/web-bluetooth/#dfn-parsing-the-blacklist
+pub fn validate_blacklist(blacklist: &str) -> Option<String> {
+    let mut result = HashMap::<&str, &str>::new();
+    for (index, line) in blacklist.lines().enumerate() {
+        let line_num = index + 1;
+        if line.is_empty() || line.starts_with("#") {
+            // Comment or blank line.
+            continue;
+        }
+        let tokens: Vec<&str> = line.split(' ').collect();
+        match tokens.len() {
+            0 => unreachable!("split(' ') never returns an empty array"),
+            1 | 2 => {
+                let uuid = tokens[0];
+                if !valid_uuid(uuid) {
+                    return Some(format!("line {}: '{}' is not a valid UUID", line_num, uuid));
+                }
+                let mut exclusion = "exclude";
+                if tokens.len() == 2 {
+                    exclusion = tokens[1];
+                    match exclusion {
+                        "exclude-reads" | "exclude-writes" => (),
+                        _ => return Some(format!(
+			    "line {}: '{}' should be 'exclude-reads', or 'exclude-writes'",
+			    line_num, exclusion)),
+                    }
+                }
+                if let Some(_) = result.insert(uuid, exclusion) {
+                    return Some(format!(
+                        "line {}: '{}' appears multiple times",
+                        line_num, uuid));
+                }
+            },
+            _ => return Some(format!("line {}: Too many tokens", line_num)),
+        }
+    }
+    return None;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs::File;
+    use std::io::prelude::*;
+
+    #[test]
+    fn test_valid_uuid() {
+	assert!( valid_uuid("00000000-0000-0000-0000-000000000000"));
+	assert!( valid_uuid("01234567-89ab-cdef-0123-456789abcdef"));
+	assert!(!valid_uuid("01234567-89AB-CDEF-0123-456789ABCDEF"));
+	assert!(!valid_uuid("g1234567-89ab-cdef-0123-456789abcdef"));
+	assert!(!valid_uuid("01234567-89ab-cdef-0123-456789abcdef0"));
+	assert!(!valid_uuid("0123456789abcdef0123456789abcdef"));
+	assert!(!valid_uuid("01234567089ab0cdef001230456789abcdef"));
+    }
+
+    #[test]
+    fn test_validate_blacklist() {
+        assert_eq!(None, validate_blacklist(""));
+        // Lines are terminated by \n, not \r\n.
+        assert_eq!(Some("line 1: '\r' is not a valid UUID".to_string()),
+                   validate_blacklist("\r\n"));
+        assert_eq!(None, validate_blacklist("# comment"));
+        assert_eq!(None, validate_blacklist("# comment
```

Comment:
The explicit \n, as used below, seem easier to read in this file. Unless you feel strongly that the literal newline included in the string is valuable I think it would be best to use the escaped version.